### PR TITLE
Gaps-NULL vs empty("") vs null vs ~ vs None vs null(Java)

### LINKED ISSUE #758: Gaps-NULL vs empty("") vs null vs ~ vs None vs null(Java)
Here are the gaps regarding the expected and actual runtime behavior.

[Actual behavior regarding empty and null inputs - GitHub.xlsx](https://github.com/CloudSlang/cloud-slang/files/338762/Actual.behavior.regarding.empty.and.null.inputs.-.GitHub.xlsx)

We have the following cases:

**_1. CS Flow Input Required - has a default value - Empty String - The execution continues with empty value_**

**ACTUAL BEHAVIOR:** 
 
Currently if a CS flow has an input required and the input has a default value, in case of empty value given, the execution will continue with empty value.

**EXPECTED BEHAVIOR:** 
 
The expected runtime behavior in this situation is that the execution should continue with default value assigned.

**HOW-TO-REPRODUCE:** 
 
1. Create a flow with: - required input
                               - default value for the input
2. Run the flow
 
**Actual behaviour:** 
Continues with empty value
 
**Expected behaviour:**
Continues with default value assigned

**_2. CS Flow Input Optional - has a default value - When empty value is given, the execution continues with empty value_**

**ACTUAL BEHAVIOR:** 
 
Currently if a CS flow has an optional input and the input has a default value, in case of empty value given, the execution will continue with empty value.

**EXPECTED BEHAVIOR:** 
 
The expected runtime behavior in this situation is that the execution should continue with default value assigned.

**HOW-TO-REPRODUCE:** 
 
1. Create a flow with: - optional input
                               - default value for the input
2. Run the flow
 
**Actual behaviour:** 
Continues with empty value
 
**Expected behaviour:**
Continues with default value assigned

**_3. CS Step Input Required - has a default - Empty Value - The execution continues with empty value assigned_**

**ACTUAL BEHAVIOR:** 
 
Currently if a CS step has an input required and the input has a default value, in case of empty value given, the execution will continue with empty value.

**EXPECTED BEHAVIOR:** 
 
The expected runtime behavior in this situation is that the execution should continue with default value assigned.

**HOW-TO-REPRODUCE:** 
 
1. Create a flow with that use an operation with: - required input
                                                                     - default value for the input
2. Run the flow
 
**Actual behaviour:** 
For operation/step input = required with default value & flow input= not required => flow continues with empty value
 
For operation/step input = required with default value & flow input= required => execution fails 
 
**Expected behaviour:**
Continues with default value assigned

**_4. CS Step Input Optional - has a default value - Empty string - Execution continues with empty value_**

**ACTUAL BEHAVIOR:** 
 
Currently if a CS step has an input optional and the input has a default value, in case of empty value given, the execution will continue with empty value.

**EXPECTED BEHAVIOR:** 
 
The expected runtime behavior in this situation is that the execution should continue with default value assigned.

**HOW-TO-REPRODUCE:** 
 
1. Create a flow with that use an operation with: - optional input
                                                                     - default value for the input
2. Run the flow
 
**Actual behaviour:** 
For operation/step input = optional with default value & flow input= not required => flow continues with empty value
 
For operation/step input = optional with default value & flow input= required => execution fails
 
**Expected behaviour:**
Continues with default value assigned

**ENVIRONMENT**
 
CloudSlang 0.9.59.4

### PR DESCRIPTION
**Pull request by gituser173**
Fixes #758
Signed-off-by: Claudiu Ban claudiu-laurian.ban@hpe.com

### CODE REVIEW

File: cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/bindings/InputsBinding.java [hunk 1]
```diff
@@ -87,7 +87,7 @@ private Value resolveValue(Input input, Map<String, ? extends Value> context, Ma
             value = ValueFactory.create(valueFromContext, sensitive);
         }
 
-        if (value == null || value.get() == null) {
+        if (value == null || value.get() == null || value.get().equals("")) {
```
Comment (Bonczidai) [inline cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/bindings/InputsBinding.java]: IMHO you should extract this logic used in two places as a method ~ `isDefined`

**Comment by Bonczidai** [comment]
.. in the meantime please also merge from master to resolve conflicts

**Comment by gituser173** [comment]
resolved conflicts

**Comment by Bonczidai** [comment]
:+1:

**Pull request merged**